Fix/Separate deploying contract and NNS updating transaction

carpawell opened the pull request:
Deploy contract in one transaction and register NNS record in another to
prevent GAS limit exceeding due to the fact that registration always takes
10+ GAS.

Signed-off-by: Pavel Karpy <carpawell@nspcc.ru>

Closes #1675.

fyrchik commented:
NNS has `SetPrice` method, can we do `SetPrice` the same trick as in https://github.com/nspcc-dev/neofs-node/pull/1608 ?

carpawell commented:
> NNS has SetPrice method, can we do SetPrice the same trick as in https://github.com/nspcc-dev/neofs-node/pull/1608 ?

I did like it was done for NeoFS contracts (deploying in [one stage](https://github.com/nspcc-dev/neofs-node/blob/89cba2cc7becdaadc55eb4d2996153318df5ddf6/cmd/neofs-adm/internal/modules/morph/initialize.go#L74), setting NNS in [another](https://github.com/nspcc-dev/neofs-node/blob/89cba2cc7becdaadc55eb4d2996153318df5ddf6/cmd/neofs-adm/internal/modules/morph/initialize.go#L94))

Also, if Im not mistaken, @realloc was against changing the price for that purpose.

realloc commented:
I didn't like the idea of changing NNS pricing in general for all.

I have nothing against temporary NNS price change, contract deploy and setting price back in the same Tx. This doesn't mean we don't have to check deploy actions sanity first =)

fyrchik commented:
Yes, the whole purpose of the optimisation is to do it temporarily, so that there are no visible side-effects for other transactions.

carpawell commented:
@fyrchik, added hack with changing price but kept separate deploying and registering. Also, added `--force` that allows deploying contract without group/with a group that differs from `group.neofs`.

AnnaShaleva commented on cmd/neofs-adm/internal/modules/morph/deploy.go:
```diff
@@ -109,24 +110,44 @@ func deployContractCmd(cmd *cobra.Command, args []string) error {
+	}
+
+	// add record to NNS
+	w = io.NewBufBinWriter()
```
BufBinWriter can be reused after previous transaction. Use `w.Reset()` to reset it.

carpawell replied:
fixed

AnnaShaleva commented on cmd/neofs-adm/internal/modules/morph/deploy.go:
```diff
@@ -95,9 +98,32 @@ func deployContractCmd(cmd *cobra.Command, args []string) error {
+
+		if !pub.Equal(group) && !force {
+			return fmt.Errorf(
+				"`%s` does not equal NNS's `%s`; use `--%s` flag to deploy with that key anyway",
+				contractWalletFilename, morphclient.NNSGroupKeyName, forceFlag,
```
Maybe some "%s doesn't contain NNS's group key..." will be better? The first argument is wallet filename, this message looks strange a bit.

carpawell replied:
decided not to deploy custom contracts with `group.neofs` group, removed that code

AnnaShaleva commented on cmd/neofs-adm/internal/modules/morph/deploy.go:
```diff
@@ -95,9 +98,32 @@ func deployContractCmd(cmd *cobra.Command, args []string) error {
+		if err != nil {
+			return fmt.Errorf("can't sign manifest group: %v", err)
+		}
+
```
Linter is unhappy.

carpawell replied:
forget that the linter already works

carpawell commented:
Removed `--force` flag and do deploy contract with the neofs group.

fyrchik commented:
@carpawell why did you separate deploy and register? `setPrice` should be enough.

carpawell commented:
> why did you separate deploy and register?

@fyrchik that was the original idea of the PR. Yes, now `SetPrice` fixes the problem. Do you want to keep registration and deploying in the same TX?

fyrchik commented:
Yes, the less transactions we have, the fewer places where something could go wrong.

carpawell commented:
@fyrchik, ok, dropped that commit.
